# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

I predict that the Random Forests Classifier will perform better since it creates more decision trees

## Split the Data into Training and Testing Sets

In [3]:
# Create the labels set `y` and features DataFrame `X`
X = data.drop('spam', axis=1)
y = data['spam']

In [4]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
data['spam'].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)
X_train

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1053            0.29               0.58           0.58          0.00   
935             0.00               0.00           0.00         40.13   
4105            0.00               0.00           0.00          0.00   
1523            0.49               0.21           0.56          0.00   
1464            0.32               0.16           0.56          0.00   
...              ...                ...            ...           ...   
1973            0.08               0.00           0.22          0.00   
3558            0.00               0.00           0.00          0.00   
3543            0.10               0.00           0.10          0.00   
1015            0.00               0.72           1.81          0.00   
1570            0.00               0.00           1.07          0.00   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1053           0.87            0.00              0.58                0.00   
935            0.00            0.00              0.00                0.00   
4105           0.79            0.79              0.00                0.00   
1523           0.28            0.21              0.00                0.28   
1464           0.32            0.23              0.04                1.24   
...             ...             ...               ...                 ...   
1973           0.04            0.00              0.00                0.00   
3558           0.19            0.00              0.00                0.00   
3543           0.00            0.00              0.00                0.00   
1015           0.00            0.36              0.00                0.36   
1570           3.22            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1053             0.00            1.16  ...                   0.0        0.000   
935              0.00            0.32  ...                   0.0        0.000   
4105             0.00            0.00  ...                   0.0        0.000   
1523             0.28            0.98  ...                   0.0        0.000   
1464             0.40            0.40  ...                   0.0        0.019   
...               ...             ...  ...                   ...          ...   
1973             0.44            0.04  ...                   0.0        0.163   
3558             0.00            0.00  ...                   0.0        0.068   
3543             0.92            0.00  ...                   0.0        0.014   
1015             0.72            1.08  ...                   0.0        0.000   
1570             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1053        0.054        0.000        0.000        0.271        0.000   
935         0.185        0.000        0.185        0.061        0.000   
4105        0.000        0.000        0.000        0.000        0.000   
1523        0.108        0.000        0.768        0.312        0.000   
1464        0.052        0.065        0.413        0.164        0.000   
...           ...          ...          ...          ...          ...   
1973        0.197        0.058        0.000        0.011        0.000   
3558        0.113        0.000        0.022        0.045        0.000   
3543        0.154        0.028        0.000        0.028        0.000   
1015        0.334        0.000        1.203        0.467        0.066   
1570        0.000        0.000        2.395        0.598        0.998   

      capital_run_length_average  capital_run_length_longest  \
1053                       1.670                          14   
935                       10.585                         124   
4105                       1.076                           2   
1523                       3.401                          94   
1464                       3.533              

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [6]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [7]:
# Fit the Standard Scaler with the training data
scaler = StandardScaler().fit(X_train)


In [8]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_train_scaled

array([[ 0.62334413,  0.26469298,  0.58248252, ..., -0.10739713,
        -0.18404341, -0.19741722],
       [-0.34759839, -0.16954656, -0.56190527, ...,  0.14635393,
         0.32955666,  0.23499022],
       [-0.34759839, -0.16954656, -0.56190527, ..., -0.12430438,
        -0.24007251, -0.41044148],
       ...,
       [-0.01279063, -0.16954656, -0.36459703, ..., -0.07566046,
        -0.09999976,  0.35104075],
       [-0.34759839,  0.36950943,  3.00937386, ...,  0.36879526,
         1.58554228,  0.71508966],
       [-0.34759839, -0.16954656,  1.54929289, ...,  2.18618224,
         1.12797131,  0.06806823]])

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [13]:
# Create and fit a Logistic Regression model
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression()
lr_model.fit(X_train, y_train)


/Users/xfigu/.conda/envs/dev/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [14]:
# Make and save testing predictions with the trained Logistic Regression model using the test data
testing_predictions = lr_model.predict(X_test)

# Review the predictions
results_df = pd.DataFrame({"Prediction": testing_predictions, "Actual": y_test})

results_df


Prediction  Actual
4495           0       0
859            1       1
818            1       1
3535           0       0
2136           0       0
...          ...     ...
2341           0       0
1507           1       1
1537           1       1
2759           0       0
1901           0       0

[1151 rows x 2 columns]

In [15]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)

0.9131190269331017

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [12]:
# Create and fit a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(random_state=1)
rf_model.fit(X_train, y_train)



RandomForestClassifier(random_state=1)

In [16]:
# Make and save testing predictions with the trained Random Forest Classifier model using the test data
rf_model_predictions = rf_model.predict(X_test)


# Review the predictions
print(rf_model_predictions)

[0 1 1 ... 1 0 0]


In [17]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test, rf_model_predictions)


print(f"Accuracy Score : {acc_score}")

Accuracy Score : 0.9617723718505647


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The Random Forest Classifier performed better than the Linear Regression model. Compared to my prediction, it turns out my prediction was correct. Random Forest Classifier had a higher accuracy score with 96.1% where as the Linear Regression Model had an accuracy score of 91.3%. 